File: app/server/services/backup.ts
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import { spawn } from 'child_process';
import config from '../utils/config';
import { safePath, formatBytes, isAllowedPath } from '../utils/validation';
import { BackupInfo, BackupResult } from '../types';

const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);
const mkdir = promisify(fs.mkdir);
const copyFile = promisify(fs.copyFile);
const unlink = promisify(fs.unlink);
const rmdir = promisify(fs.rmdir);

export const BACKUP_BASE_DIR = config.backup.baseDir;

async function ensureDir(dir: string): Promise<void> {
    try {
        await mkdir(dir, { recursive: true });
    } catch (e) {
        if ((e as NodeJS.ErrnoException).code !== 'EEXIST') throw e;
    }
}

function isSafeBackupBase(dir: string): boolean {
    const normalized = safePath(dir);
    if (!normalized) return false;
    return normalized.startsWith('/vol') && normalized.includes('/@appshare');
}

function isLogFile(filename: string): boolean {
    const lower = filename.toLowerCase();
    if (lower.endsWith('.log')) return true;
    if (lower.includes('.log.')) return true;
    if (lower.includes('log') && lower.endsWith('.txt')) return true;
    return false;
}

interface CollectedFile {
    fullPath: string;
    relativePath: string;
}

async function collectLogFiles(dir: string, baseDir: string, maxFiles: number = 500, maxFileSize: number = config.backup.maxFileSizeBytes): Promise<CollectedFile[]> {
    const files: CollectedFile[] = [];

    async function walk(currentDir: string): Promise<void> {
        if (files.length >= maxFiles) return;

        try {
            const entries = await readdir(currentDir, { withFileTypes: true });
            for (const entry of entries) {
                if (files.length >= maxFiles) break;

                const fullPath = path.join(currentDir, entry.name);

                if (entry.isDirectory()) {
                    await walk(fullPath);
                } else if (entry.isFile() && isLogFile(entry.name)) {
                    try {
                        const stats = await stat(fullPath);
                        if (stats.size > maxFileSize) continue;
                    } catch {
                        continue;
                    }
                    const relativePath = path.relative(baseDir, fullPath);
                    files.push({ fullPath, relativePath });
                }
            }
        } catch {
            // 忽略权限错误
        }
    }

    await walk(dir);
    return files;
}

async function copyFileToBackup(src: string, dest: string): Promise<void> {
    const destDir = path.dirname(dest);
    await ensureDir(destDir);
    await copyFile(src, dest);
}

async function createTarGz(sourceDir: string, outputFile: string): Promise<void> {
    const safeSourceDir = safePath(sourceDir);
    const safeOutputFile = safePath(outputFile);

    if (!safeSourceDir || !safeOutputFile) {
        throw new Error('无效的路径');
    }

    const baseName = path.basename(safeSourceDir);
    const parentDir = path.dirname(safeSourceDir);

    if (baseName.includes('\0') || parentDir.includes('\0')) {
        throw new Error('路径包含非法字符');
    }

    return new Promise((resolve, reject) => {
        const proc = spawn('tar', [
            '-czf',
            safeOutputFile,
            '-C', parentDir,
            baseName
        ], {
            timeout: 300000
        });

        let stderr = '';

        proc.stderr.on('data', (data) => {
            stderr += data.toString();
        });

        proc.on('close', (code) => {
            if (code === 0) {
                resolve();
            } else {
                reject(new Error(`tar 命令失败: ${stderr || `退出码 ${code}`}`));
            }
        });

        proc.on('error', (err) => {
            reject(new Error(`tar 命令执行失败: ${err.message}`));
        });
    });
}

async function removeDir(dir: string): Promise<void> {
    const safeDir = safePath(dir);
    if (!safeDir) return;

    const dangerousPaths = ['/', '/var', '/etc', '/usr', '/bin', '/sbin', '/lib', '/boot'];
    if (dangerousPaths.includes(safeDir)) {
        throw new Error('不允许删除系统目录');
    }

    try {
        const stats = await stat(safeDir);
        if (!stats.isDirectory()) return;
    } catch {
        return;
    }

    async function rmRecursive(currentDir: string): Promise<void> {
        const entries = await readdir(currentDir, { withFileTypes: true });

        for (const entry of entries) {
            const fullPath = path.join(currentDir, entry.name);
            if (entry.isDirectory()) {
                await rmRecursive(fullPath);
            } else {
                await unlink(fullPath);
            }
        }

        await rmdir(currentDir);
    }

    await rmRecursive(safeDir);
}

export async function performBackup(): Promise<BackupResult> {
    if (!isSafeBackupBase(BACKUP_BASE_DIR)) {
        return {
            backupPath: '',
            files: 0,
            backupSize: '0B',
            errors: ['备份目录不安全']
        };
    }
    const timestamp = new Date().toISOString()
        .replace(/[:.]/g, '-')
        .slice(0, 19)
        .replace(/[^0-9T\-]/g, '');

    const backupDir = path.join(BACKUP_BASE_DIR, `backup-${timestamp}`);
    const backupFile = `${backupDir}.tar.gz`;

    const results: BackupResult = {
        backupPath: backupFile,
        files: 0,
        backupSize: '0B',
        errors: []
    };

    try {
        await ensureDir(BACKUP_BASE_DIR);
        await ensureDir(backupDir);

        let totalBytes = 0;

        for (const logDir of config.logDirs) {
            const normalizedDir = safePath(logDir);
            if (!normalizedDir) continue;
            if (!isAllowedPath(normalizedDir, config.logDirs)) continue;

            try {
                const stats = await stat(normalizedDir).catch(() => null);
                if (!stats || !stats.isDirectory()) continue;

                const dirName = path.basename(normalizedDir);
                const targetDir = path.join(backupDir, dirName);

                const files = await collectLogFiles(normalizedDir, normalizedDir, config.backup.maxFiles, config.backup.maxFileSizeBytes);

                for (const file of files) {
                    try {
                        const srcStats = await stat(file.fullPath);
                        if (totalBytes + srcStats.size > config.backup.maxTotalBytes) {
                            results.errors.push('备份内容超过大小限制');
                            break;
                        }
                        const targetPath = path.join(targetDir, file.relativePath);
                        await copyFileToBackup(file.fullPath, targetPath);
                        results.files++;
                        totalBytes += srcStats.size;
                    } catch (e) {
                        results.errors.push(`复制失败: ${file.fullPath} - ${(e as Error).message}`);
                    }
                }

                if (totalBytes >= config.backup.maxTotalBytes) {
                    break;
                }
            } catch (e) {
                results.errors.push(`处理目录失败: ${logDir} - ${(e as Error).message}`);
            }
        }

        if (results.files > 0) {
            await createTarGz(backupDir, backupFile);

            await removeDir(backupDir);

            const stats = await stat(backupFile);
            if (stats.size > config.backup.maxTotalBytes) {
                await unlink(backupFile);
                throw new Error('备份文件超过大小限制');
            }
            results.backupSize = formatBytes(stats.size);
        } else {
            await removeDir(backupDir);
            results.errors.push('没有找到日志文件');
        }
    } catch (e) {
        results.errors.push(`备份失败: ${(e as Error).message}`);
    }

    return results;
}

export async function listBackups(): Promise<BackupInfo[]> {
    const backups: BackupInfo[] = [];

    try {
        await stat(BACKUP_BASE_DIR).catch(() => null);
        const entries = await readdir(BACKUP_BASE_DIR, { withFileTypes: true }).catch(() => []);

        for (const entry of entries) {
            if (!entry.isFile() || !entry.name.endsWith('.tar.gz')) continue;

            const fullPath = path.join(BACKUP_BASE_DIR, entry.name);
            try {
                const stats = await stat(fullPath);
                backups.push({
                    name: entry.name,
                    path: fullPath,
                    size: formatBytes(stats.size),
                    created: stats.mtime
                });
            } catch {
                // 忽略无法访问的文件
            }
        }
    } catch {
        // 忽略错误
    }

    backups.sort((a, b) => b.created.getTime() - a.created.getTime());
    return backups;
}

export async function deleteBackup(backupPath: string): Promise<void> {
    const safeBackupPath = safePath(backupPath);

    if (!safeBackupPath) {
        throw new Error('无效的路径');
    }

    if (!safeBackupPath.startsWith(BACKUP_BASE_DIR)) {
        throw new Error('只能删除备份目录下的文件');
    }

    if (!safeBackupPath.endsWith('.tar.gz')) {
        throw new Error('只能删除备份文件');
    }

    await unlink(safeBackupPath);
}

export async function cleanOldBackups(keepDays: number = 30): Promise<number> {
    const backups = await listBackups();
    const cutoff = Date.now() - keepDays * 24 * 60 * 60 * 1000;
    let deleted = 0;

    for (const backup of backups) {
        if (backup.created.getTime() < cutoff) {
            try {
                await deleteBackup(backup.path);
                deleted++;
            } catch {
                // 忽略错误
            }
        }
    }

    return deleted;
}

File: app/server/internal/services/backup.go
package services

import (
	"fmt"
	"io"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/sushazhi/fnos-logmanager/internal/config"
	"github.com/sushazhi/fnos-logmanager/internal/types"
	"github.com/sushazhi/fnos-logmanager/internal/utils"
)

const backupBaseDir = "" // set dynamically from config

// ensureBackupDir ensures the backup directory exists.
func ensureBackupDir(dir string) error {
	return os.MkdirAll(dir, 0755)
}

// isSafeBackupBase checks if the backup base directory is safe.
func isSafeBackupBase(dir string) bool {
	normalized := utils.SafePath(dir)
	if normalized == "" {
		return false
	}
	if !strings.HasPrefix(normalized, "/") {
		return false
	}
	for _, dangerous := range []string{"/etc", "/proc", "/sys"} {
		if strings.HasPrefix(normalized, dangerous) {
			return false
		}
	}
	return true
}

// isLogFile checks if a filename looks like a log file.
func isLogFile(filename string) bool {
	lower := strings.ToLower(filename)
	if strings.HasSuffix(lower, ".log") {
		return true
	}
	if strings.Contains(lower, ".log.") {
		return true
	}
	if strings.Contains(lower, "log") && strings.HasSuffix(lower, ".txt") {
		return true
	}
	return false
}

type collectedFile struct {
	FullPath     string
	RelativePath string
}

// collectLogFiles recursively collects log files from a directory.
func collectLogFiles(dir, baseDir string, maxFiles int, maxFileSize int64) ([]collectedFile, error) {
	var files []collectedFile

	err := filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil // skip permission errors
		}
		if len(files) >= maxFiles {
			return io.EOF // stop walk
		}
		if d.IsDir() {
			return nil
		}
		if !isLogFile(d.Name()) {
			return nil
		}

		info, err := d.Info()
		if err != nil {
			return nil
		}
		if info.Size() > maxFileSize {
			return nil
		}

		relPath, err := filepath.Rel(baseDir, path)
		if err != nil {
			return nil
		}
		files = append(files, collectedFile{FullPath: path, RelativePath: relPath})
		return nil
	})
	if err == io.EOF {
		err = nil
	}
	return files, err
}

// copyFileToBackup copies a file to the backup directory.
func copyFileToBackup(src, dest string) error {
	destDir := filepath.Dir(dest)
	if err := ensureBackupDir(destDir); err != nil {
		return err
	}

	srcFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer srcFile.Close()

	destFile, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer destFile.Close()

	_, err = io.Copy(destFile, srcFile)
	return err
}

// createTarGz creates a tar.gz archive from a source directory.
func createTarGz(sourceDir, outputFile string) error {
	safeSourceDir := utils.SafePath(sourceDir)
	safeOutputFile := utils.SafePath(outputFile)
	if safeSourceDir == "" || safeOutputFile == "" {
		return fmt.Errorf("无效的路径")
	}

	baseName := filepath.Base(safeSourceDir)
	parentDir := filepath.Dir(safeSourceDir)

	cmd := exec.Command("tar", "-czf", safeOutputFile, "-C", parentDir, baseName)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("tar 命令失败: %s: %w", string(output), err)
	}
	return nil
}

// removeDir safely removes a directory tree.
func removeDir(dir string) error {
	safeDir := utils.SafePath(dir)
	if safeDir == "" {
		return nil
	}

	dangerousPaths := []string{"/", "/var", "/etc", "/usr", "/bin", "/sbin", "/lib", "/boot"}
	for _, d := range dangerousPaths {
		if safeDir == d {
			return fmt.Errorf("不允许删除系统目录")
		}
	}

	return os.RemoveAll(safeDir)
}

// PerformBackup creates a full backup of all log directories.
func PerformBackup() types.BackupResult {
	cfg := config.Get()
	backupDir := cfg.Backup.BaseDir

	if !isSafeBackupBase(backupDir) {
		return types.BackupResult{
			BackupPath: "",
			Files:      0,
			BackupSize: "0B",
			Errors:     []string{"备份目录不安全"},
		}
	}

	timestamp := time.Now().Format("2006-01-02T15-04-05")
	tmpBackupDir := filepath.Join(backupDir, "backup-"+timestamp)
	backupFile := tmpBackupDir + ".tar.gz"

	result := types.BackupResult{
		BackupPath: backupFile,
		Files:      0,
		BackupSize: "0B",
		Errors:     []string{},
	}

	if err := ensureBackupDir(backupDir); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("创建备份目录失败: %v", err))
		return result
	}
	if err := ensureBackupDir(tmpBackupDir); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("创建临时目录失败: %v", err))
		return result
	}

	var totalBytes int64

	for _, logDir := range cfg.LogDirs {
		normalizedDir := utils.SafePath(logDir)
		if normalizedDir == "" {
			continue
		}

		stat, err := os.Stat(normalizedDir)
		if err != nil || !stat.IsDir() {
			continue
		}

		dirName := filepath.Base(normalizedDir)
		targetDir := filepath.Join(tmpBackupDir, dirName)

		files, err := collectLogFiles(normalizedDir, normalizedDir, cfg.Backup.MaxFiles, cfg.Backup.MaxFileSizeBytes)
		if err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("收集文件失败 %s: %v", logDir, err))
			continue
		}

		for _, f := range files {
			if totalBytes >= cfg.Backup.MaxTotalBytes {
				result.Errors = append(result.Errors, "备份内容超过大小限制")
				break
			}

			info, err := os.Stat(f.FullPath)
			if err != nil {
				continue
			}
			if totalBytes+info.Size() > cfg.Backup.MaxTotalBytes {
				result.Errors = append(result.Errors, "备份内容超过大小限制")
				break
			}

			targetPath := filepath.Join(targetDir, f.RelativePath)
			if err := copyFileToBackup(f.FullPath, targetPath); err != nil {
				result.Errors = append(result.Errors, fmt.Sprintf("复制失败: %s - %v", f.FullPath, err))
				continue
			}
			result.Files++
			totalBytes += info.Size()
		}

		if totalBytes >= cfg.Backup.MaxTotalBytes {
			break
		}
	}

	if result.Files > 0 {
		if err := createTarGz(tmpBackupDir, backupFile); err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("创建归档失败: %v", err))
			return result
		}

		if err := removeDir(tmpBackupDir); err != nil {
			slog.Warn("清理临时目录失败", "error", err)
		}

		stat, err := os.Stat(backupFile)
		if err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("检查备份文件失败: %v", err))
			return result
		}
		if stat.Size() > cfg.Backup.MaxTotalBytes {
			os.Remove(backupFile)
			result.Errors = append(result.Errors, "备份文件超过大小限制")
			return result
		}
		result.BackupSize = utils.FormatBytes(stat.Size())
	} else {
		removeDir(tmpBackupDir)
		result.Errors = append(result.Errors, "没有找到日志文件")
	}

	return result
}

// ListBackups returns a list of existing backup files.
func ListBackups() []types.BackupInfo {
	cfg := config.Get()
	backupDir := cfg.Backup.BaseDir

	var backups []types.BackupInfo

	entries, err := os.ReadDir(backupDir)
	if err != nil {
		return backups
	}

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".tar.gz") {
			continue
		}

		fullPath := filepath.Join(backupDir, entry.Name())
		info, err := entry.Info()
		if err != nil {
			continue
		}

		backups = append(backups, types.BackupInfo{
			Name:    entry.Name(),
			Path:    fullPath,
			Size:    utils.FormatBytes(info.Size()),
			Created: info.ModTime(),
		})
	}

	sort.Slice(backups, func(i, j int) bool {
		return backups[i].Created.After(backups[j].Created)
	})

	return backups
}

// DeleteBackup deletes a backup file.
func DeleteBackup(backupPath string) error {
	safePath := utils.SafePath(backupPath)
	if safePath == "" {
		return fmt.Errorf("无效的路径")
	}

	cfg := config.Get()
	if !strings.HasPrefix(safePath, cfg.Backup.BaseDir) {
		return fmt.Errorf("只能删除备份目录下的文件")
	}

	if !strings.HasSuffix(safePath, ".tar.gz") {
		return fmt.Errorf("只能删除备份文件")
	}

	return os.Remove(safePath)
}

// CleanOldBackups removes backups older than keepDays.
func CleanOldBackups(keepDays int) (int, error) {
	backups := ListBackups()
	cutoff := time.Now().AddDate(0, 0, -keepDays)
	deleted := 0

	for _, b := range backups {
		if b.Created.Before(cutoff) {
			if err := DeleteBackup(b.Path); err != nil {
				slog.Warn("删除旧备份失败", "path", b.Path, "error", err)
				continue
			}
			deleted++
		}
	}
	return deleted, nil
}
